### PR TITLE
Update combat related feature

### PR DESCRIPTION
**Pull request by tytannial**
Fix file name and add MagicAnimation table pointer.
Add H3CombatCreature::ApplyAfterHitSpell function pointer.
Add H3CombatCreature::ApplyAfterHitAbility function pointer.
Fix H3CombatManager::ReportDamageDone function param.
Add H3CombatManager::ApplyMagicDamage function pointer.
Add H3CombatManager::ApplyAnimationToLastHitArmy function pointer.
Add H3SoundManager::PlaySoundByFileAsync function pointer.

### CODE REVIEW

**Comment by tytannial** [comment]
done.

File: include/h3api/H3Combat/H3MagicAnimation.hpp [hunk 1]
```diff
@@ -14,16 +14,20 @@
...
-	struct H3MagicAnimation
-	{
-		LPCSTR defName;
-		LPCSTR name;
-		INT32 type;
-	};
+    struct H3MagicAnimation
+    {
+        _H3API_SIZE_(0x0C);
+        _H3API_VTABLE_(0x641E18);
```
Comment (RoseKavalier) [inline include/h3api/H3Combat/H3MagicAnimation.hpp]: `_H3API_GET_INFO_(H3MagicAnimation, 0x43E500+3)`
`_H3API_VTABLE_` is for virtual tables.

File: include/h3api/H3Managers/H3CombatManager.cpp [hunk 2]
```diff
@@ -97,9 +101,9 @@ namespace h3
 	{
 		THISCALL_5(VOID, 0x4963C0, this, id, target, timeStep, show_hit);
 	}
-	_H3API_ VOID H3CombatManager::ReportDamageDone(H3Spell* spell, LPCSTR attackerName, INT32 damageDone, H3CombatCreature* target, INT32 killedCount)
+    _H3API_ VOID H3CombatManager::ReportDamageDone(LPCSTR attackerName, BOOL showLog, INT32 damageDone, H3CombatCreature* target, INT32 killedCount)
```
Comment (RoseKavalier) [inline include/h3api/H3Managers/H3CombatManager.cpp]: BOOL showLog -> INT32 numAttackers

File: include/h3api/H3Managers/H3CombatManager.hpp [hunk 3]
```diff
@@ -304,9 +311,31 @@ namespace h3
...
+         * @param damage Basic damage
+         * @param spellId Which spell
+         * @param atkHero Attacker hero
+         * @param defHero Defender hero
+         * @param target Target creature
+         * @param trueCast If true, it will cast real spell to target and generate combat log.
+         *                 when false caculate damage but no effect.
+         * @return Damage would be done
+        */
+        _H3API_ INT32 ApplyMagicDamage(INT32 damage, INT32 spellId, H3Hero* atkHero, H3Hero* defHero, H3CombatCreature* target, BOOL trueCast);
```
Comment (RoseKavalier) [inline include/h3api/H3Managers/H3CombatManager.hpp]: This function does not apply magic damage, it applies modifiers to your damage.
`trueCast` is incorrect.

File: include/h3api/H3Managers/H3SoundManager.hpp [hunk 4]
```diff
@@ -57,6 +57,13 @@ namespace h3
 		 * @param milliseconds The maximum duration in milliseconds, -1 defaults to 10,000ms
 		*/
 		_H3API_ VOID PlaySoundAsync(LPCSTR wav_name, INT32 duration = -1);
+        /**
+         * @brief play a sound directly.
+         * 
+         * @param wav_name The name of the file to play
+         * @return Played Wav file, WARNNING Homm3HD MOD will modify this return pointer to INT32 value as waiting timesteps
+         */
+        _H3API_ static H3WavFile* PlaySoundByFileAsync(LPCSTR wav_name);
```
Comment (RoseKavalier) [inline include/h3api/H3Managers/H3SoundManager.hpp]: Wrong return type

**Review by RoseKavalier — CHANGES_REQUESTED** [review]
See comments! Sorry don't have much time these days.

File: include/h3api/H3Managers/H3SoundManager.hpp (hunk 4 above)
Comment (tytannial) [reply include/h3api/H3Managers/H3SoundManager.hpp]: > See comments! Sorry don't have much time these days.

investigating

File: include/h3api/H3Managers/H3CombatManager.hpp (hunk 3 above)
Comment (tytannial) [reply include/h3api/H3Managers/H3CombatManager.hpp]: comment in wog source.....wrong one +1...

File: include/h3api/H3Managers/H3SoundManager.hpp [hunk 5]
```diff
@@ -57,6 +57,13 @@ namespace h3
 		 * @param milliseconds The maximum duration in milliseconds, -1 defaults to 10,000ms
 		*/
 		_H3API_ VOID PlaySoundAsync(LPCSTR wav_name, INT32 duration = -1);
+        /**
+         * @brief play a sound directly.
+         * 
+         * @param wav_name The name of the file to play
+         * @return Played Wav file, WARNNING Homm3HD MOD will modify this return pointer to INT32 value as waiting timesteps
+         */
+        _H3API_ static INT64 PlaySoundByFileAsync(LPCSTR wav_name);
```
Comment (RoseKavalier) [inline include/h3api/H3Managers/H3SoundManager.hpp:66]: It's not INT64 either but at least this is the correct size (:

**Review by RoseKavalier — APPROVED** [review]

File: include/h3api/H3Managers/H3SoundManager.hpp (hunk 5 above)
Comment (tytannial) [reply include/h3api/H3Managers/H3SoundManager.hpp:66]: > It's not INT64 either but at least this is the correct size (:

Maybe a float, i guess. 😭

**Pull request merged**